File: src/types/report.py
"""Pydantic models for engine results and research reports."""

from datetime import datetime

from pydantic import BaseModel

from src.types.enums import EngineType, ResearchStatus
from src.types.events import CompletenessAudit, DeepDive, ViralEvent


class EngineResult(BaseModel):
    """Result from a single research engine run."""

    engine: EngineType
    status: ResearchStatus
    raw_markdown: str
    tldr: str | None
    viral_events: list[ViralEvent]
    deep_dives: list[DeepDive]
    completeness_audit: CompletenessAudit | None
    started_at: datetime
    completed_at: datetime
    duration_seconds: float
    error_message: str | None


class ResearchReport(BaseModel):
    """A daily research report combining both engine results."""

    report_id: str
    run_date: datetime
    gemini_result: EngineResult | None
    langchain_result: EngineResult | None
    created_at: datetime

File: rust-backend/src/types/report.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use super::enums::ResearchStatus;
use super::events::{CompletenessAudit, DeepDive, ViralEvent};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EngineResult {
    pub status: ResearchStatus,
    pub raw_markdown: String,
    pub tldr: Option<String>,
    pub viral_events: Vec<ViralEvent>,
    pub deep_dives: Vec<DeepDive>,
    pub completeness_audit: Option<CompletenessAudit>,
    pub started_at: DateTime<Utc>,
    pub completed_at: DateTime<Utc>,
    pub duration_seconds: f64,
    pub error_message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ResearchReport {
    pub report_id: String,
    pub run_date: DateTime<Utc>,
    pub result: Option<EngineResult>,
    pub created_at: DateTime<Utc>,
}
